File: macos-app-management/Sources/MacOSAppRuntimeCore/RuntimeStateMachine.swift
public enum RuntimeTransitionError: Error, Equatable, Sendable {
    case negativeTimestamp(Int64)
    case nonMonotonicTimestamp(previous: Int64, received: Int64)
}

public struct RuntimeStateMachine: Sendable {
    public private(set) var state: RuntimeState

    public init(runtimeSessionID: String) {
        self.state = RuntimeState(runtimeSessionID: runtimeSessionID)
    }

    public mutating func apply(_ fact: RuntimeFact) throws -> [UsageSegment] {
        guard fact.observedAtMs >= 0 else {
            throw RuntimeTransitionError.negativeTimestamp(fact.observedAtMs)
        }

        if let previous = state.lastObservedAtMs, fact.observedAtMs < previous {
            throw RuntimeTransitionError.nonMonotonicTimestamp(
                previous: previous,
                received: fact.observedAtMs
            )
        }

        var next = state
        let endReason = apply(fact.kind, to: &next)
        let isCheckpoint = isSnapshot(fact.kind)
        let applicationChanged = hasApplicationIdentityChanged(
            from: state.application,
            to: next.application
        )

        var completed: [UsageSegment] = []
        if let open = state.openSegment,
           isCheckpoint || applicationChanged || !next.isEligibleForUsage
        {
            if fact.observedAtMs > open.startAtMs {
                completed.append(
                    UsageSegment(
                        id: segmentID(
                            runtimeSessionID: state.runtimeSessionID,
                            ordinal: state.nextSegmentOrdinal
                        ),
                        runtimeSessionID: state.runtimeSessionID,
                        application: open.application,
                        startAtMs: open.startAtMs,
                        endAtMs: fact.observedAtMs,
                        durationMilliseconds: fact.observedAtMs - open.startAtMs,
                        endReason: endReason
                    )
                )
                next.nextSegmentOrdinal += 1
            }
            next.openSegment = nil
        }

        if next.openSegment == nil,
           next.isEligibleForUsage,
           let application = next.application
        {
            next.openSegment = RuntimeState.OpenSegment(
                application: application,
                startAtMs: fact.observedAtMs
            )
        }

        next.lastObservedAtMs = fact.observedAtMs
        state = next
        return completed
    }

    private func apply(_ kind: RuntimeFact.Kind, to state: inout RuntimeState) -> SegmentEndReason {
        switch kind {
        case let .applicationActivated(application):
            state.application = application
            return .applicationSwitch

        case let .userActivityChanged(activity):
            state.userActivity = activity
            return activity == .idle ? .userIdle : .stateCorrection

        case let .sessionChanged(session):
            state.sessionState = session
            return session == .active ? .stateCorrection : .sessionUnavailable

        case let .powerChanged(power):
            state.powerState = power
            return power == .asleep ? .systemSleep : .stateCorrection

        case let .snapshot(snapshot):
            state.application = snapshot.application
            state.userActivity = snapshot.userActivity
            state.sessionState = snapshot.sessionState
            state.powerState = snapshot.powerState
            return .periodicSnapshot
        }
    }

    private func isSnapshot(_ kind: RuntimeFact.Kind) -> Bool {
        if case .snapshot = kind {
            return true
        }
        return false
    }

    private func hasApplicationIdentityChanged(
        from previous: ApplicationIdentity?,
        to current: ApplicationIdentity?
    ) -> Bool {
        previous?.runtimeIdentity != current?.runtimeIdentity
    }

    private func segmentID(runtimeSessionID: String, ordinal: UInt64) -> String {
        "\(runtimeSessionID):\(ordinal)"
    }
}

File: app-runtime-management/agents/windows/src/TimeOnChrome.AppRuntime.Core/RuntimeStateMachine.cs
using System.Globalization;

namespace TimeOnChrome.AppRuntime.Core;

public sealed class RuntimeTransitionException : Exception
{
    private RuntimeTransitionException(
        string code,
        long? value = null,
        long? previous = null,
        long? received = null)
        : base(code)
    {
        Code = code;
        Value = value;
        Previous = previous;
        Received = received;
    }

    public string Code { get; }

    public long? Value { get; }

    public long? Previous { get; }

    public long? Received { get; }

    internal static RuntimeTransitionException NegativeTimestamp(long value) =>
        new("negativeTimestamp", value: value);

    internal static RuntimeTransitionException NonMonotonicTimestamp(long previous, long received) =>
        new("nonMonotonicTimestamp", previous: previous, received: received);
}

public sealed class RuntimeStateMachine
{
    public RuntimeStateMachine(string runtimeSessionID)
    {
        State = new RuntimeState { RuntimeSessionID = runtimeSessionID };
    }

    public RuntimeState State { get; private set; }

    public IReadOnlyList<UsageSegment> Apply(RuntimeFact fact)
    {
        ArgumentNullException.ThrowIfNull(fact);

        if (fact.ObservedAtMs < 0)
        {
            throw RuntimeTransitionException.NegativeTimestamp(fact.ObservedAtMs);
        }

        if (State.LastObservedAtMs is long previous && fact.ObservedAtMs < previous)
        {
            throw RuntimeTransitionException.NonMonotonicTimestamp(previous, fact.ObservedAtMs);
        }

        var next = State;
        var endReason = ApplyFact(fact, ref next);
        var isCheckpoint = fact.Kind == RuntimeFactKind.Snapshot;
        var applicationChanged = HasApplicationIdentityChanged(State.Application, next.Application);
        var completed = new List<UsageSegment>(capacity: 1);

        if (State.OpenSegment is { } open
            && (isCheckpoint || applicationChanged || !next.IsEligibleForUsage))
        {
            if (fact.ObservedAtMs > open.StartAtMs)
            {
                completed.Add(new UsageSegment(
                    Id: string.Concat(
                        State.RuntimeSessionID,
                        ":",
                        State.NextSegmentOrdinal.ToString(CultureInfo.InvariantCulture)),
                    RuntimeSessionID: State.RuntimeSessionID,
                    Application: open.Application,
                    StartAtMs: open.StartAtMs,
                    EndAtMs: fact.ObservedAtMs,
                    DurationMilliseconds: fact.ObservedAtMs - open.StartAtMs,
                    EndReason: endReason));
                next = next with { NextSegmentOrdinal = next.NextSegmentOrdinal + 1 };
            }

            next = next with { OpenSegment = null };
        }

        if (next.OpenSegment is null
            && next.IsEligibleForUsage
            && next.Application is { } application)
        {
            next = next with
            {
                OpenSegment = new OpenUsageSegment(application, fact.ObservedAtMs),
            };
        }

        State = next with { LastObservedAtMs = fact.ObservedAtMs };
        return completed;
    }

    private static SegmentEndReason ApplyFact(RuntimeFact fact, ref RuntimeState state)
    {
        switch (fact.Kind)
        {
            case RuntimeFactKind.ApplicationActivated:
                state = state with { Application = fact.Application };
                return SegmentEndReason.ApplicationSwitch;

            case RuntimeFactKind.UserActivityChanged:
                var activity = Require(fact.UserActivity, nameof(fact.UserActivity));
                state = state with { UserActivity = activity };
                return activity == UserActivityState.Idle
                    ? SegmentEndReason.UserIdle
                    : SegmentEndReason.StateCorrection;

            case RuntimeFactKind.SessionChanged:
                var session = Require(fact.SessionState, nameof(fact.SessionState));
                state = state with { SessionState = session };
                return session == UserSessionState.Active
                    ? SegmentEndReason.StateCorrection
                    : SegmentEndReason.SessionUnavailable;

            case RuntimeFactKind.PowerChanged:
                var power = Require(fact.PowerState, nameof(fact.PowerState));
                state = state with { PowerState = power };
                return power == SystemPowerState.Asleep
                    ? SegmentEndReason.SystemSleep
                    : SegmentEndReason.StateCorrection;

            case RuntimeFactKind.Snapshot:
                var snapshot = fact.Snapshot
                    ?? throw new InvalidOperationException("Snapshot fact requires snapshot payload.");
                state = state with
                {
                    Application = snapshot.Application,
                    UserActivity = snapshot.UserActivity,
                    SessionState = snapshot.SessionState,
                    PowerState = snapshot.PowerState,
                };
                return SegmentEndReason.PeriodicSnapshot;

            default:
                throw new InvalidOperationException($"Unsupported runtime fact kind: {fact.Kind}.");
        }
    }

    private static T Require<T>(T? value, string propertyName)
        where T : struct =>
        value ?? throw new InvalidOperationException($"Runtime fact requires {propertyName} payload.");

    private static bool HasApplicationIdentityChanged(
        ApplicationIdentity? previous,
        ApplicationIdentity? current) =>
        previous?.Platform != current?.Platform
        || !string.Equals(
            previous?.RuntimeIdentity,
            current?.RuntimeIdentity,
            StringComparison.Ordinal);
}
